File: smallest_difference/smallest_difference.py
def smallest_difference(arr1, arr2):
    # Sort both the array
    arr1.sort()
    arr2.sort()

    # setting up var to point to the elements of the array
    idx1 = 0      
    idx2 = 0
    output = []

    # Initialize to infinity ie. the biggest value
    smallest_diff = float("inf")

    while (idx1 < len(arr1) and idx2 < len(arr2)):
        # checking the difference b/w current elements of array
        diff = abs(arr1[idx1] - arr2[idx2])

        if diff == 0:
            # means the numbers are same from both the array and ther diff is 0
            return [arr1[idx1], arr2[idx2]], diff

        # checking the current diff wih the smallest diff and updating the
        # output and diff values
        if diff < smallest_diff:
            smallest_diff = diff
            output = [arr1[idx1], arr2[idx2]]

        # Moving the index pointer
        if arr1[idx1] < arr2[idx2]:
            idx1 += 1;
        else:
            idx2 += 1;

    return output, smallest_diff





# Execution begins here
if __name__ == "__main__":
    arr1 = [-1, 5, 10, 20, 28, 3];
    arr2 = [26, 134, 135, 15, 17];

    output, diff = smallest_difference(arr1, arr2)
    print("NUmber pair with smallest difference: ", output)
    print("Smallest difference: ", diff)

File: smallest_difference/smallest_difference.cpp
#include <iostream>
#include <vector>
#include <bits/stdc++.h>
#include <cstdlib>

using namespace std;

void smallest_difference(vector<int>& arr1, vector<int>& arr2, vector<int>& out) {
   // Sort both the array
   sort(arr1.begin(), arr1.end());
   sort(arr2.begin(), arr2.end());

   // var to point to the array elements
   int idx1 = 0;
   int idx2 = 0;

   // Using the "out" array last element for holding the smallest diff value
   // initialize smallest difference with biggest value as a starter value
   out[2] = 99999;

   while (idx1 < arr1.size() && idx2 < arr2.size()) {
      // Difference b/w current array elements
      int diff = abs(arr1[idx1] - arr2[idx2]);

      if (diff == 0) {
         // if the difference is zero, then these are the pairs with 
         // smallest difference
         out[0] = arr1[idx1];
         out[1] = arr2[idx2];
         out[2] = diff;

         return;
      }

      // validating current diff with smallest_diff and updating the value
      if (diff < out[2]) {
         out[0] = arr1[idx1];
         out[1] = arr2[idx2];         
         out[2] = diff;
      }

      // Moving the array position pointers
      if (arr1[idx1] < arr2[idx2])
         idx1++;
      else 
         idx2++;
   }

   return;
}





/* Execution begins here */
int main() {
   // INPUT
   vector<int> arr1 = {-1, 5, 10, 20, 28, 3};
   vector<int> arr2 = {26, 134, 135, 15, 17};

   // OUTPUT [pair1, pair2, smallest_diff]
   vector<int> out(3);

   smallest_difference(arr1, arr2, out);

   cout << "PAIRS: " << out[0] << ", " << out[1] << endl;
   cout << "And their difference is " << out[2] << endl;

   return 0;
}
